core/src/main/java/io/undertow/util/Cookies.java

```diff
@@ -211,51 +212,80 @@ public static Map<String, Cookie> parseRequestCookies(int maxCookies, boolean al
         return parseRequestCookies(maxCookies, allowEqualInValue, cookies, LegacyCookieSupport.COMMA_IS_SEPARATOR);
     }
 
+    @Deprecated(since="2.3.1", forRemoval=true)
     public static void parseRequestCookies(int maxCookies, boolean allowEqualInValue, List<String> cookies, Set<Cookie> parsedCookies) {
         parseRequestCookies(maxCookies, allowEqualInValue, cookies, parsedCookies, LegacyCookieSupport.COMMA_IS_SEPARATOR);
     }
 
+    public static void parseRequestCookies(int maxCookies, boolean allowEqualInValue, List<String> cookies, MultiValueHashListStorage<String, Cookie> parsedCookies) {
+        parseRequestCookies(maxCookies, allowEqualInValue, cookies, parsedCookies, LegacyCookieSupport.COMMA_IS_SEPARATOR);
+    }
+
     @Deprecated
     static Map<String, Cookie> parseRequestCookies(int maxCookies, boolean allowEqualInValue, List<String> cookies, boolean commaIsSeperator) {
         return parseRequestCookies(maxCookies, allowEqualInValue, cookies, commaIsSeperator, LegacyCookieSupport.ALLOW_HTTP_SEPARATORS_IN_V0);
     }
 
+    @Deprecated(since="2.3.1", forRemoval=true)
     static void parseRequestCookies(int maxCookies, boolean allowEqualInValue, List<String> cookies, Set<Cookie> parsedCookies, boolean commaIsSeperator) {
         parseRequestCookies(maxCookies, allowEqualInValue, cookies, parsedCookies, commaIsSeperator, LegacyCookieSupport.ALLOW_HTTP_SEPARATORS_IN_V0);
     }
 
+    static void parseRequestCookies(int maxCookies, boolean allowEqualInValue, List<String> cookies, MultiValueHashListStorage<String, Cookie> parsedCookies, boolean commaIsSeperator) {
+        parseRequestCookies(maxCookies, allowEqualInValue, cookies, parsedCookies, commaIsSeperator, LegacyCookieSupport.ALLOW_HTTP_SEPARATORS_IN_V0);
+    }
+
+    @Deprecated(since="2.3.1", forRemoval=true)
     static Map<String, Cookie> parseRequestCookies(int maxCookies, boolean allowEqualInValue, List<String> cookies, boolean commaIsSeperator, boolean allowHttpSepartorsV0) {
         if (cookies == null) {
             return new TreeMap<>();
         }
-        final Set<Cookie> parsedCookies = new HashSet<>();
+        final MultiValueHashListStorage<String, Cookie> parsedCookies = new MultiValueHashListStorage<>();
         for (String cookie : cookies) {
-            parseCookie(cookie, parsedCookies, maxCookies, allowEqualInValue, commaIsSeperator, allowHttpSepartorsV0);
+            parseCookies(cookie, parsedCookies, maxCookies, allowEqualInValue, commaIsSeperator, allowHttpSepartorsV0);
         }
 
         final Map<String, Cookie> retVal = new TreeMap<>();
-        for (Cookie cookie : parsedCookies) {
-            retVal.put(cookie.getName(), cookie);
+        Iterator<Cookie> cookiesIterator = parsedCookies.valuesIterator();
+        Cookie cookie = null;
+        for (;cookiesIterator.hasNext(); ) {
```

Comment:
This is not a problem at all but 'while' would be a better choice than 'for' with empty  'initialization' and 'post action' section. Another option is to move 'Cookie cookie = null' declaration to 'for initialization' section.

Reply:
I have no idea "why" I did it this way